### PR TITLE
Adding Config options for Siteadmins to configure whether clients can use Chat/autocomplete/commands

### PR DESCRIPTION
**Pull request by arafatkatze**
Main Issue https://github.com/sourcegraph/sourcegraph/issues/58631
Related PR https://github.com/sourcegraph/dev-private/pull/82

## Test plan
Run in Dev mode and test the GQL by tweaking the site admin config.

### CODE REVIEW

File: internal/conf/computed.go [hunk 1]
```diff
@@ -793,6 +793,31 @@ func GetCompletionsConfig(siteConfig schema.SiteConfiguration) (c *conftypes.Com
 	return computedConfig
 }
 
+func GetConfigFeatures(siteConfig schema.SiteConfiguration) (c *conftypes.ConfigFeatures) {
+	// If cody is disabled, don't use any of the other features.
+	if !codyEnabled(siteConfig) {
+		return nil
+	}
+	configFeatures := siteConfig.ConfigFeatures
+	// If no completions configuration is set at all, but cody is enabled, assume a default configuration
```
Comment (cbart) [inline internal/conf/computed.go:802]: I think you meant features configuration not completions configuration.

File: internal/conf/computed.go [hunk 2]
```diff
@@ -793,6 +793,31 @@ func GetCompletionsConfig(siteConfig schema.SiteConfiguration) (c *conftypes.Com
...
+	// where all the features are enabled this is to handle edge cases where no config is set etc
+	if configFeatures == nil {
+		defaultConfig := &conftypes.ConfigFeatures{
+			Chat:         true,
+			AutoComplete: true,
+			Commands:     true,
+		}
+		return defaultConfig
```
Comment (cbart) [inline internal/conf/computed.go:810]:
```suggestion
		return &conftypes.ConfigFeatures{
			Chat:         true,
			AutoComplete: true,
			Commands:     true,
		}
```

nit

File: cmd/frontend/graphqlbackend/site.go
Comment (cbart) [inline cmd/frontend/graphqlbackend/site.go:1]: Can you add a unit test where you change site config, query graphql and see the changes reflected?

Comment (arafatkatze) [reply cmd/frontend/graphqlbackend/site.go:1]: Added some testing: in a standard way to test these conftypes. Hopefully this is okay.

[Another example](https://github.com/sourcegraph/sourcegraph/blob/994286815e94d74deacf00819c4e8e9e5bf7eb40/internal/conf/computed_test.go#L317)

**Review by cbart — COMMENTED** [review]
I just saw that I did not click submit in review! Apologies for the lag.

**Review by slimsag — APPROVED** [review]
This looks good to me as-is. It immediately lets us test / work with the client-side changes in #2559 so feel free to merge this.

As [discussed in Slack](https://sourcegraph.slack.com/archives/C05JDP433DL/p1704408478471069), I will send a follow-up PR soon (in the next day or so) to replace this approach with a RBAC-based approach instead which has a nicer UI and other benefits.

For now, though, this site config approach is great - so I say go ahead :)

File: schema/site.schema.json [hunk 3]
```diff
@@ -2866,6 +2866,29 @@
           "perUserDailyLimit": 100
         }
       ]
+    },
+    "configFeatures": {
```
Comment (eseliger) [inline schema/site.schema.json:2870]: drive-by comment: 

The name `configFeatures` isn't very expressive, could we name this something like `cody.features` or so, and what's the relationship between this setting and `completions.enabled` and `cody.enabled` for example?

Comment (arafatkatze) [reply schema/site.schema.json:2870]: @eseliger There are other places with deprecated placeholders for enabling/disabling completions/autocomplete and I could have probably used them but I wanted as simple and easy place to just make a clean GQL query to be able to test things.

I do agree that this is not full+ best naming and also misses some good tests like @cbart  suggested but merging this now just as a placeholder for my testing and we shall remove it on the coming the RBAC stuff.

File: internal/conf/computed_test.go
[the remark this answers is not in the document]
Comment (arafatkatze) [reply internal/conf/computed_test.go:1]: @cbart Great suggestions Cezary I went and tried to implement them and leant a few things on the way but then I learnt for Stephen that we are going to replace all of this anyway for RBAC stuff so I am merging this now just for testing. Hopefully thats okay with you?

File: schema/site.schema.json (hunk 3 above)
Comment (slimsag) [reply schema/site.schema.json:2870]: Ack! I plan to remove this entirely and replace it with RBAC instead anyway, this is just helpful in the interim for testing the client-side changes primarily - so doesn't need to be perfect. I'll cc you on the new approach PR
